x/stake/validator.go

```diff
@@ -0,0 +1,242 @@
+	DelegatorShares sdk.Rat    `json:"delegator_shares"` // total shares issued to a validator's delegators
+
+	Description        Description `json:"description"`            // description terms for the validator
+	BondHeight         int64       `json:"validator_bond_height"`  // earliest height as a bonded validator
+	BondIntraTxCounter int16       `json:"validator_bond_counter"` // block-local tx index of validator change
+	ProposerRewardPool sdk.Coins   `json:"proposer_reward_pool"`   // XXX reward pool collected from being the proposer
+
+	Commission            sdk.Rat `json:"commission"`              // XXX the commission rate of fees charged to any delegators
+	CommissionMax         sdk.Rat `json:"commission_max"`          // XXX maximum commission rate which this validator can ever charge
+	CommissionChangeRate  sdk.Rat `json:"commission_change_rate"`  // XXX maximum daily increase of the validator commission
+	CommissionChangeToday sdk.Rat `json:"commission_change_today"` // XXX commission rate change today, reset each day (UTC time)
+
+	// fee related
+	PrevBondedShares sdk.Rat `json:"prev_bonded_shares"` // total shares of a global hold pools
+}
+
+// Validators - list of Validators
+type Validators []Validator
+
+// NewValidator - initialize a new validator
+func NewValidator(address sdk.Address, pubKey crypto.PubKey, description Description) Validator {
+	return Validator{
+		Status:                sdk.Unbonded,
+		Address:               address,
+		PubKey:                pubKey,
+		PShares:               NewUnbondedShares(sdk.ZeroRat()),
+		DelegatorShares:       sdk.ZeroRat(),
+		Description:           description,
+		BondHeight:            int64(0),
+		BondIntraTxCounter:    int16(0),
+		ProposerRewardPool:    sdk.Coins{},
+		Commission:            sdk.ZeroRat(),
+		CommissionMax:         sdk.ZeroRat(),
+		CommissionChangeRate:  sdk.ZeroRat(),
+		CommissionChangeToday: sdk.ZeroRat(),
+		PrevBondedShares:      sdk.ZeroRat(),
+	}
+}
+
+// only the vitals - does not check bond height of IntraTxCounter
+func (v Validator) equal(c2 Validator) bool {
+	return v.Status == c2.Status &&
+		v.PubKey.Equals(c2.PubKey) &&
+		bytes.Equal(v.Address, c2.Address) &&
+		v.PShares.Equal(c2.PShares) &&
+		v.DelegatorShares.Equal(c2.DelegatorShares) &&
+		v.Description == c2.Description &&
+		//v.BondHeight == c2.BondHeight &&
+		//v.BondIntraTxCounter == c2.BondIntraTxCounter && // counter is always changing
+		v.ProposerRewardPool.IsEqual(c2.ProposerRewardPool) &&
+		v.Commission.Equal(c2.Commission) &&
+		v.CommissionMax.Equal(c2.CommissionMax) &&
+		v.CommissionChangeRate.Equal(c2.CommissionChangeRate) &&
+		v.CommissionChangeToday.Equal(c2.CommissionChangeToday) &&
+		v.PrevBondedShares.Equal(c2.PrevBondedShares)
+}
+
+// intended to be used with require/assert:  require.True(ValEq(...))
+func ValEq(t *testing.T, exp, got Validator) (*testing.T, bool, string, Validator, Validator) {
```

Comment:
Should this be in the common testfile?

Reply:
sure, maybe a better place, moved to test_common.go